### DISCUSSION TITLE
Average value of array

### DISCUSSION THREAD

**Discussion by bflo96** [OpenModelica/OpenModelica#12400] [Q&A] [answered]
Hi everyone,

Within a model, I was trying to calculate an average value an array.

I have something like this: Major[m].Array[n].Value
I want to calculate for each **m**, the average values of **n**, so I want something like this:

```
Average[1] = (Major[1].Array[1].Value + Major[1].Array[2].Value + (...) + Major[1].Array[n].Value) / n;
Average[2] = (Major[2].Array[1].Value + Major[2].Array[2].Value + (...) + Major[2].Array[n].Value) / n;
(...)
Average[m] = (Major[m].Array[1].Value + Major[m].Array[2].Value + (...) + Major[m].Array[n].Value) / n;
```

As I want to vary **m** and **n** according to my problem, I want this to be looped.
I was trying to do this within my model directly, but it gives me more equations than it should:

```
model A
[PackageName].Major Major[m] ; //this is where Array[i].Value is defined 
Real Average[m];
parameter Integer m=5;
parameter Integer n=5;
Real aux;

for i in 1:m loop
aux= 0;
for j in 1:n loop
aux= aux+ Major[i].Array[j].Value;
end for;
Average[i] = aux;
end for;
end A;
```
 It gives 60 equations and 31 variables. When I look into the Instantiate Model, I can see that it does indeed create a new equation whenever it enters the loops:
```
aux = 0.0;
aux = aux + Array[1,1];
aux = aux + Array[1,2];
aux = aux + Array[1,3];
aux = aux + Array[1,4];
aux = aux + Array[1,5];
Average[1] = aux;
```

My second test was to create a function, which I did as well, but it gives me an error ot type mismatch for the first argument: 

> _expected type was Real[Lines,Columns] but it interprets it as Real only:_

```
function test_ave
input Real Matriz[Lines,Columns];
input Integer Lines;
input Integer Columns;
output Real Average[Lines];
protected
  Real sum;
algorithm
  for i in 1:Lines loop
    sum := 0;
    for j in 1:Columns loop
      sum := sum + Matriz[i,j];
    end for;
    Average[i] := sum / Columns;
  end for;
end test_ave;
```

What am I doing wrong?

Thank you in advance.

_EDIT: Corrected mistaked mentioned in the replies_

**Comment by casella** [comment]
I don't understand the definition of Major, is it an array?

**Reply by bflo96** [reply]
I'm not sure if my wording is correct but it's a model I created and I'm instantiating it with a certain dimension (it repeats with different attributes and I want to simulate all cases at once). Not sure if I can call it an array... 😅

**Reply by casella** [reply]
But are you instantiating Major or A? Or both? I miss the full picture. Anyway, I guess the problem is equations vs. algorithms

**Reply by bflo96** [reply]
Oh you're right of course. As I was trying to put here an over-simplified version of my real code, I wrote it with a mistake. 
I want to instantiate Major within the model A.
I'm going to edit the original to avoid further confusion in case anyone checks this discussion 😊

**Comment by casella** [comment, marked as answer]
In any case, the way you compute the average is an algorithm, I guess if you write it like this it should work:
```
model A
extends Major; // this is where Major[i].Array[i].Value is defined 
Real Average[m];
parameter Integer m=5;
parameter Integer n=5;
Real aux;
algorithm
for i in 1:m loop
aux:= 0;
for j in 1:n loop
aux := aux+ Major[i].Array[j].Value;
end for;
Average[i] := aux;
end for;
end A;
```
Otherwise, you have multiple assignment equations to the same aux variable, so of course you system has more equations than variables 😃

**Reply by bflo96** [reply]
Thank you! I just have another question: how should I organize the model if I need both algoritm and equation? Is there any specific rule (like one that should come first)?

**Comment by perost** [comment]
A possibly simpler solution would be to not use any loops and instead use the builtin sum operator, something like this:
```Modelica
  Average = {sum(Major[i].Array.Value) / size(Major[i].Array, 1) for i}; 
```
As for solving it with a function I can't say from the information you've given why it didn't work for you, but a more canonical way of implementing it would be:
```Modelica
function test_ave
  input Real Matriz[:,:];
  output Real Average[Lines];
protected
  parameter Integer Lines = size(Matriz, 1);
  parameter Integer Columns = size(Matriz, 2); 
protected
  Real sum;
algorithm
  for i in 1:Lines loop
    sum := 0;
    for j in 1:Columns loop
      sum := sum + Matriz[i,j];
    end for;
    Average[i] := sum / Columns;
  end for;
end test_ave;
```
I.e. just use the `size` operator to get the size inside the function instead of requiring the size to be passed as an argument. And then just call the function as `test_ave(Major.Array.Value)`.

Also, the common naming convention in Modelica is to start class names with upper case and variable names with lower case. It might help to avoid some confusion since most Modelica users would at first glance assume that e.g. `Major` is some kind of class, but it's not a strict rule so up to you if you want to follow it or not.

**Reply by bflo96** [reply]
@perost Your suggestion came right on time, as I was just reading about the builtin sum operator. It solves my problem perfectly 😊

I'll look into the function you sent, as it could be useful in the future and I want to know as well why mine didn't work. I'll let you know if I manage to implement it.

Also, thank you for the heads up regarding the naming convention. I understand my code might still be confusing sometimes and your suggestion will definitely make it easier for me to explain my doubts better.

**Comment by casella** [comment]
In any case, a model can have several algorithm sections and algorithm section, each starting with the corresponding keyword. Each algorithm section is conceptually handled as as single function call: the outputs are the variables appearing on the left-hand-side of the assigments, the inputs are all the other variables showing up in the right-hand-sides.
